### PR TITLE
Wire hybrid bus

### PR DESCRIPTION
**Pull request by RmbRT**
Enables perun-network/perun-websocket-backend#2.

### CODE REVIEW

File: wire/hybridbus.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	"context"
+	"fmt"
+
+	"polycry.pt/poly-go/errors"
+	"polycry.pt/poly-go/sync/atomic"
+)
+
+var _ Bus = (*hybridBus)(nil)
+
+// A Bus that can send messages over a local bus and a remote bus.
```
Comment (cryptphil) [inline wire/hybridbus.go:27]:
```suggestion
// A Bus that uses multiple buses when trying to send a message.
// Could be used with a local bus and a remote bus.
```

File: wire/hybridbus.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+func (b *hybridBus) Publish(ctx context.Context, e *Envelope) error {
+	// Whether an attempt to publish the envelope succeeded.
+	var success atomic.Bool
+	// Sub-context for sending, cancel on first success.
+	sending, cancelSending := context.WithCancel(ctx)
+	defer cancelSending()
+
+	errg := errors.NewGatherer()
+	for _, bus := range b.buses {
+		bus := bus
```
Comment (cryptphil) [inline wire/hybridbus.go:60]: Maybe add a comment on why this is needed? Otherwise, this looks like a mistake.

File: wire/hybridbus.go (hunk 1 above)
Comment (matthiasgeihs) [reply wire/hybridbus.go:27]: I prefer @RmbRT 's version. It's clearer in describing what this object provides. Your suggestion focuses on how it's implemented, which is not so relevant here.

File: wire/hybridbus.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+type hybridBus struct {
+	buses []Bus
+}
+
+// NewHybridBus creates a hybrid bus that sends and receives messages over
+// multiple buses. Publishing a message over a hybrid bus will attempt to
+// publish it over all its sub-buses simultaneously. Subscribing a client to the
+// hybrid bus will subscribe it to all its sub-buses. The sub-buses can still be
+// accessed independently for more fine-grained control.
+func NewHybridBus(buses ...Bus) Bus {
```
Comment (matthiasgeihs) [inline wire/hybridbus.go:37]: This only works for a certain kind of bus, right? It should be documented here which properties a bus must fulfill in order to be used with the hybrid bus.

File: wire/hybridbus.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+			if err == nil {
+				success.Set()
+				cancelSending()
+			}
+			return err
+		})
+	}
+
+	// Wait until all sending attempts terminated. This is when either the
+	// context expires, all buses fail, or at least one bus succeeds.
```
Comment (matthiasgeihs) [inline wire/hybridbus.go:74]: This information would be valuable as part of the function documentation.

File: wire/hybridbus_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	. "perun.network/go-perun/wire"
+	"perun.network/go-perun/wire/test"
+)
+
+func TestHybridBus_New(t *testing.T) {
```
Comment (matthiasgeihs) [inline wire/hybridbus_test.go:26]: I am not the biggest fan of such tests. They are quite tedious to maintain and don't bring much value in my opinion.

File: wire/net/bus_test.go [hunk 6]
```diff
@@ -19,7 +19,7 @@ import (
 
 	"github.com/stretchr/testify/assert"
 
-	"perun.network/go-perun/wire"
+	. "perun.network/go-perun/wire"
```
Comment (matthiasgeihs) [inline wire/net/bus_test.go:22]: We usually don't do this in go-perun. Is there a specific reason you use it here?

Comment (RmbRT) [reply wire/net/bus_test.go:22]: it keeps the diff small.

File: wire/hybridbus.go (hunk 2 above)
Comment (RmbRT) [reply wire/hybridbus.go:60]: Not going to comment this, as it is required and common in go. If you call an anonymous function in a goroutine inside a for loop, then all variables are referenced by name, not by value. This means that the for loop iterator variable will change before or during the execution of the goroutine. A local copy is created to pin the value the loop variable had during that iteration.
If we started commenting this, there'd be no end to the comments we would have to make.

File: wire/net/bus_test.go (hunk 6 above)
Comment (RmbRT) [reply wire/net/bus_test.go]: This was a lazy fix, I think I had copied some code from inside the wire package, and it took less typing to fix the type names that way. I re-did it now.

File: wire/hybridbus.go (hunk 3 above)
Comment (RmbRT) [reply wire/hybridbus.go:44]: Added description to the constructor.

**Review by matthiasgeihs — APPROVED** [review]

**Pull request merged**